core/app.go

```diff
@@ -0,0 +1,342 @@
+	if 1 == os.Getpid() {
+		ReapChildren()
+	}
+	a.HandleSignals()
+	a.HandlePolling()
+
+	if len(flag.Args()) != 0 {
+		// Run our main application and capture its stdout/stderr.
+		// This will block until the main application exits and then os.Exit
+		// with the exit code of that application.
+		a.Command = utils.ArgsToCmd(flag.Args())
+		a.Command.Stderr = os.Stderr
+		a.Command.Stdout = os.Stdout
+		code, err := utils.ExecuteAndWait(a.Command)
+		if err != nil {
+			log.Println(err)
+		}
+		// Run the PostStop handler, if any, and exit if it returns an error
+		if postStopCode, err := utils.RunWithFields(a.PostStopCmd, log.Fields{"process": "PostStop"}); err != nil {
+			os.Exit(postStopCode)
+		}
+		os.Exit(code)
+	}
+
+	// block forever, as we're polling in the two polling functions and
+	// did not os.Exit by waiting on an external application.
+	select {}
+}
+
+// ToggleMaintenanceMode marks all services for maintenance
+func (a *App) ToggleMaintenanceMode() {
+	a.maintModeLock.RLock()
+	a.signalLock.Lock()
+	defer a.signalLock.Unlock()
+	defer a.maintModeLock.RUnlock()
+	a.paused = !a.paused
+	if a.paused {
+		a.forAllServices(markServiceForMaintenance)
+	}
+}
+
+// InMaintenanceMode checks if the App is in maintenance mode
+func (a *App) InMaintenanceMode() bool {
+	// we wrap access to `paused` in a RLock so that if we're in the middle of
+	// marking services for maintenance we don't get stale reads
+	a.maintModeLock.RLock()
+	defer a.maintModeLock.RUnlock()
+	return a.paused
+}
+
+// Terminate kills the application
+func (a *App) Terminate() {
+	a.signalLock.Lock()
+	defer a.signalLock.Unlock()
+	a.stopPolling()
+	a.forAllServices(deregisterService)
+
+	// Run and wait for preStop command to exit
+	utils.RunWithFields(a.PreStopCmd, log.Fields{"process": "PreStop"})
+
+	cmd := a.Command
+	if cmd == nil || cmd.Process == nil {
+		// Not managing the process, so don't do anything
+		return
+	}
+	if a.StopTimeout > 0 {
+		if err := cmd.Process.Signal(syscall.SIGTERM); err != nil {
+			log.Warnf("Error sending SIGTERM to application: %s", err)
+		} else {
+			time.AfterFunc(time.Duration(a.StopTimeout)*time.Second, func() {
+				log.Infof("Killing Process %#v", cmd.Process)
+				cmd.Process.Kill()
+			})
+			return
+		}
+	}
+	log.Infof("Killing Process %#v", a.Command.Process)
+	cmd.Process.Kill()
+}
+
+func (a *App) stopPolling() {
+	for _, quit := range a.QuitChannels {
+		quit <- true
+	}
+}
+
+func markServiceForMaintenance(service *services.Service) {
+	log.Infof("Marking for maintenance: %s", service.Name)
+	service.MarkForMaintenance()
+}
+
+func deregisterService(service *services.Service) {
+	log.Infof("Deregistering service: %s", service.Name)
+	service.Deregister()
+}
+
+// Reload ...
+func (a *App) Reload() error {
+	a.signalLock.Lock()
+	defer a.signalLock.Unlock()
+	log.Infof("Reloading configuration.")
+
+	newCfg, err := config.ParseConfig(a.ConfigFlag)
+	if err != nil {
```

Comment:
We never log this error, which means we fail to reload silently. We should consider doing `log.Errorf` here.